Bug: `SimpleArrayMixinCalculators` needs to handle non-c-contiguous memory buffer

If the array is not c-contiguous, the return values of the functions in `class SimpleArrayMixinCalculators;` are wrong.

In other words, it is necessary to fix the function `SimpleArrayMixinCalculators::min()`, `SimpleArrayMixinCalculators::max()`, `SimpleArrayMixinCalculators::sum()`, `SimpleArrayMixinCalculators::abs()`.

For example:

Unittest:
```python3
    def test_sum2(self):

        nparr = np.arange(24, dtype='float64').reshape((2, 3, 4))
        nparr = nparr[:,:,::2] ## Slicing
        sarr = modmesh.SimpleArrayFloat64(array=nparr) 
        npsum = nparr.sum()
        ssum = sarr.sum()
        self.assertEqual(npsum, ssum)
```

Source code of `sum()`:
```c++
    value_type sum() const
    {
        value_type initial;
        if constexpr (is_complex_v<value_type>)
        {
            initial = value_type();
        }
        else
        {
            initial = 0;
        }

        auto athis = static_cast<A const *>(this);
        if constexpr (!std::is_same_v<bool, std::remove_const_t<value_type>>)
        {
            for (size_t i = 0; i < athis->size(); ++i)
            {
                initial += athis->data(i);
            }
        }
        else
        {
            for (size_t i = 0; i < athis->size(); ++i)
            {
                initial |= athis->data(i);
            }
        }
        return initial;
    }
```

Console log:
```console

self = <test_buffer.SimpleArrayCalculatorsTC testMethod=test_sum2>

    def test_sum2(self):
    
        nparr = np.arange(24, dtype='float64').reshape((2, 3, 4))
        nparr = nparr[:,:,::2]
        sarr = modmesh.SimpleArrayFloat64(array=nparr)
        npsum = nparr.sum()
        ssum = sarr.sum()
>       self.assertEqual(npsum, ssum)
E       AssertionError: np.float64(132.0) != 66.0

tests/test_buffer.py:962: AssertionError
================================================================================================================================================================================== short test summary info ==================================================================================================================================================================================
FAILED tests/test_buffer.py::SimpleArrayCalculatorsTC::test_sum2 - AssertionError: np.float64(132.0) != 66.0
```

--- a/cpp/modmesh/buffer/SimpleArray.hpp
+++ b/cpp/modmesh/buffer/SimpleArray.hpp
@@ -152,11 +152,7 @@ public:
     value_type median() const
     {
         auto athis = static_cast<A const *>(this);
-        size_t n = 1;
-        for (size_t i = 0; i < athis->ndim(); ++i)
-        {
-            n *= athis->shape(i);
-        }
+        const size_t n = athis->size();
         small_vector<T> acopy(n);
         auto sidx = athis->first_sidx();
         size_t i = 0;
@@ -205,11 +201,7 @@ public:
     real_type var(size_t ddof) const
     {
         auto athis = static_cast<A const *>(this);
-        size_t n = 1;
-        for (size_t i = 0; i < athis->ndim(); ++i)
-        {
-            n *= athis->shape(i);
-        }
+        const size_t n = athis->size();
         if (n <= ddof)
         {
             throw std::runtime_error("SimpleArray::var(): ddof must be less than the number of elements");
@@ -723,17 +715,6 @@ public:
     {
         if (buffer)
         {
-            const size_t nbytes = ITEMSIZE *
-                                  std::accumulate(shape.begin(),
-                                                  shape.end(),
-                                                  static_cast<size_t>(1),
-                                                  std::multiplies<size_t>());
-            if (nbytes != buffer->nbytes())
-            {
-                throw std::runtime_error(Formatter() << "SimpleArray: shape byte count " << nbytes
-                                                     << " differs from buffer " << buffer->nbytes());
-            }
-
             m_shape = shape;
             m_stride = stride;
         }
@@ -834,8 +815,22 @@ public:
 
     explicit operator bool() const noexcept { return bool(m_buffer) && bool(*m_buffer); }
 
-    size_t nbytes() const noexcept { return m_buffer ? m_buffer->nbytes() : 0; }
-    size_t size() const noexcept { return nbytes() / ITEMSIZE; }
+    size_t nbytes() const noexcept { return size() * ITEMSIZE; }
+    /**
+     * In the following numpy documentaions,
+     * `size()` are calculated based on the number of the elements.
+     * https://numpy.org/doc/2.2/reference/generated/numpy.ndarray.size.html
+     * SimpleArray matchs the behavior of numpy.ndarray.
+     */
+    size_t size() const noexcept
+    {
+        size_t size = 1;
+        for (size_t it = 0; it < m_shape.size(); ++it)
+        {
+            size *= m_shape[it];
+        }
+        return size;
+    }
 
     using iterator = T *;
     using const_iterator = T const *;
@@ -1090,11 +1085,11 @@ private:
         {
             throw std::out_of_range(Formatter() << "SimpleArray: index " << it << " < -nghost: " << -static_cast<ssize_t>(m_nghost));
         }
-        if (it >= static_cast<ssize_t>(size() - m_nghost))
+        if (it >= static_cast<ssize_t>((buffer().nbytes() / ITEMSIZE) - m_nghost))
         {
             throw std::out_of_range(
-                Formatter() << "SimpleArray: index " << it << " >= " << size() - m_nghost
-                            << " (size: " << size() << " - nghost: " << m_nghost << ")");
+                Formatter() << "SimpleArray: index " << it << " >= " << (buffer().nbytes() / ITEMSIZE) - m_nghost
+                            << " (buffer size: " << (buffer().nbytes() / ITEMSIZE) << " - nghost: " << m_nghost << ")");
         }
     }
 

--- a/cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp
+++ b/cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp
@@ -78,6 +78,7 @@ class MODMESH_PYTHON_WRAPPER_VISIBILITY WrapSimpleArray
                         modmesh::detail::shape_type shape;
                         modmesh::detail::shape_type stride;
                         constexpr size_t itemsize = wrapped_type::itemsize();
+                        constexpr size_t span = 0;
                         for (ssize_t i = 0; i < arr_in.ndim(); ++i)
                         {
                             shape.push_back(arr_in.shape(i));
@@ -87,10 +88,49 @@ class MODMESH_PYTHON_WRAPPER_VISIBILITY WrapSimpleArray
                         const bool is_c_contiguous = (arr_in.flags() & py::array::c_style) == py::array::c_style;
                         const bool is_f_contiguous = (arr_in.flags() & py::array::f_style) == py::array::f_style;
 
-                        std::shared_ptr<ConcreteBuffer> const buffer = ConcreteBuffer::construct(
-                            arr_in.nbytes(),
-                            arr_in.mutable_data(),
-                            std::make_unique<ConcreteBufferNdarrayRemover>(arr_in));
+                        py::array owner = arr_in;
+                        /*
+                         * In the following document, it introduces the base object in ndarray.
+                         * https://numpy.org/doc/2.2/reference/generated/numpy.ndarray.base.html
+                         * The `array.base` is base object if memory is from some other object.
+                         * If object owns its memory, base is None.
+                         */
+                        while (true)
+                        {
+                            const py::object b = owner.attr("base");
+                            if (b.is_none() || !py::isinstance<py::array>(b))
+                            {
+                                break;
+                            }
+                            auto next = b.cast<py::array>();
+                            /*
+                             * Prevent the infinite loop.
+                             * For example, the following code will create a loop:
+                             * nparr = np.arange(24, dtype='float64').reshape((2, 3, 4))
+                             * nparr = nparr[::2, ::2, ::2]
+                             */
+                            if (next.ptr() == owner.ptr())
+                            {
+                                break;
+                            }
+                            owner = next;
+                        }
+
+                        char * base_ptr = static_cast<char *>(owner.mutable_data());
+                        char * view_ptr = static_cast<char *>(arr_in.mutable_data());
+                        const ptrdiff_t offset_bytes = view_ptr - base_ptr;
+                        if (offset_bytes < 0)
+                        {
+                            throw std::runtime_error("Unexpected negative offset!");
+                        }
+                        const size_t true_owner_nbytes = owner.nbytes();
+                        const size_t view_nbytes = true_owner_nbytes - static_cast<size_t>(offset_bytes);
+                        auto remover = std::make_unique<ConcreteBufferNdarrayRemover>(owner);
+                        const std::shared_ptr<ConcreteBuffer> buffer =
+                            ConcreteBuffer::construct(
+                                view_nbytes,
+                                view_ptr,
+                                std::move(remover));
                         return wrapped_type(shape, stride, buffer, is_c_contiguous, is_f_contiguous);
                     }),
                 py::arg("array"))

--- a/tests/test_buffer.py
+++ b/tests/test_buffer.py
@@ -320,7 +320,7 @@ class SimpleArrayBasicTC(unittest.TestCase):
 
         with self.assertRaisesRegex(
                 IndexError,
-                r"SimpleArray: index 14 >= 14 \(size: 24 - nghost: 10\)"
+                r"SimpleArray: index 14 >= 14 \(buffer size: 24 - nghost: 10\)"
         ):
             sarr[14]
 
@@ -332,7 +332,7 @@ class SimpleArrayBasicTC(unittest.TestCase):
 
         with self.assertRaisesRegex(
                 IndexError,
-                r"SimpleArray: index 14 >= 14 \(size: 24 - nghost: 10\)"
+                r"SimpleArray: index 14 >= 14 \(buffer size: 24 - nghost: 10\)"
         ):
             sarr[14] = 1
 
@@ -489,6 +489,8 @@ class SimpleArrayBasicTC(unittest.TestCase):
         np_sarr = np.empty(shape, dtype='float64')
         py_sarr = modmesh.SimpleArrayFloat64(array=np_sarr)
         self.assertTupleEqual(shape, py_sarr.shape)
+        self.assertEqual(np_sarr.nbytes, py_sarr.nbytes)
+        self.assertEqual(np_sarr.size, py_sarr.size)
 
         shape = (5, 5, 5, 5)
         np_sarr = np.empty(shape, dtype='float64')
@@ -500,6 +502,8 @@ class SimpleArrayBasicTC(unittest.TestCase):
 
         py_sarr = modmesh.SimpleArrayFloat64(array=np_sarr)
         self.assertTupleEqual(shape, py_sarr.shape)
+        self.assertEqual(np_sarr.nbytes, py_sarr.nbytes)
+        self.assertEqual(np_sarr.size, py_sarr.size)
 
     def test_SimpleArray_from_ndarray_content(self):
 
@@ -524,6 +528,8 @@ class SimpleArrayBasicTC(unittest.TestCase):
             for j in range(4):
                 for k in range(6):
                     self.assertEqual(parr[i, j, k], sarr[i, j, k])
+        self.assertEqual(parr.nbytes, sarr.nbytes)
+        self.assertEqual(parr.size, sarr.size)
 
     def test_SimpleArray_from_ndarray_transpose(self):
         ndarr = np.arange(350, dtype='float64').reshape((5, 7, 10))
@@ -1039,6 +1045,13 @@ class SimpleArrayCalculatorsTC(unittest.TestCase):
         self.assertEqual(npmed.real, smed.real)
         self.assertEqual(npmed.imag, smed.imag)
 
+        nparr = np.arange(24, dtype='float64').reshape((2, 3, 4))
+        nparr = nparr[::2, ::2, ::2]
+        sarr = modmesh.SimpleArrayFloat64(array=nparr)
+        npavg = np.median(nparr)
+        savg = sarr.median()
+        self.assertEqual(npavg, savg)
+
     def test_average(self):
         nparr = np.arange(24, dtype='float64')
         np.random.shuffle(nparr)
@@ -1061,6 +1074,13 @@ class SimpleArrayCalculatorsTC(unittest.TestCase):
         savg = sarr.average()
         self.assertEqual(npavg, savg)
 
+        nparr = np.arange(24, dtype='float64').reshape((2, 3, 4))
+        nparr = nparr[::2, ::2, ::2]
+        sarr = modmesh.SimpleArrayFloat64(array=nparr)
+        npavg = np.average(nparr)
+        savg = sarr.average()
+        self.assertEqual(npavg, savg)
+
     def test_mean(self):
         nparr = np.arange(24, dtype='float64')
         np.random.shuffle(nparr)
@@ -1078,6 +1098,13 @@ class SimpleArrayCalculatorsTC(unittest.TestCase):
 
         nparr = np.arange(81, dtype='float64').reshape((3, 3, 3, 3))
         np.random.shuffle(nparr)
+        sarr = modmesh.SimpleArrayFloat64(array=nparr)
+        npmean = np.mean(nparr)
+        smean = sarr.mean()
+        self.assertEqual(npmean, smean)
+
+        nparr = np.arange(24, dtype='float64').reshape((2, 3, 4))
+        nparr = nparr[::2, ::2, ::2]
         sarr = modmesh.SimpleArrayFloat64(array=nparr)
         npmean = np.mean(nparr)
         smean = sarr.mean()
@@ -1108,6 +1135,13 @@ class SimpleArrayCalculatorsTC(unittest.TestCase):
         svar = sarr.var()
         self.assertEqual(npvar, svar)
 
+        nparr = np.arange(24, dtype='float64').reshape((2, 3, 4))
+        nparr = nparr[::2, ::2, ::2]
+        sarr = modmesh.SimpleArrayFloat64(array=nparr)
+        npvar = np.var(nparr)
+        svar = sarr.var()
+        self.assertEqual(npvar, svar)
+
     def test_std(self):
         nparr = np.arange(24, dtype='float64')
         np.random.shuffle(nparr)
@@ -1131,6 +1165,13 @@ class SimpleArrayCalculatorsTC(unittest.TestCase):
         sarr = modmesh.SimpleArrayComplex128(array=nparr)
         npstd = np.std(nparr)
         sstd = sarr.std()
+        self.assertEqual(npstd, sstd)
+
+        nparr = np.arange(24, dtype='float64').reshape((2, 3, 4))
+        nparr = nparr[::2, ::2, ::2]
+        sarr = modmesh.SimpleArrayFloat64(array=nparr)
+        npstd = np.var(nparr)
+        sstd = sarr.var()
         self.assertEqual(npstd, sstd)
 
     def type_convertor(self, dtype):

